Make instance errors generic

swsnr opened the pull request:
Replace explicit result enums per operation type, e.g., `InstanceUpdateResult` with an `Either` over a newly introduced `InstanceError` and the old "success" type, and rename the success types to remove the now redundant "success" from their name.

Define the new `InstanceError` as ADT with specialised variants for all previously known error cases of instance operations.

Rewrite instance operations to make use of the monadic nature of `Either`, thus reducing nesting and branching to make the code more straightforward and simpler.  

In a related change explicitly specify the body parsers of all controller methods to shift the work of parsing and validating JSON to Play, and thus removing boilerplate from the controller actions, which makes all controller actions trivial wrappers around the controller operations.

In an unrelated change turn `Role` into an `enumerated` enum just because I can 😎

swsnr committed:
Add generic type class to convert data types to play HTTP results

swsnr committed:
Add generic model for instance errors

swsnr committed:
Replace InstanceDeletionResult with Either over InstanceError

swsnr committed:
Explicitly specify body parsers

Reduces complexity as Play takes over JSON parsing, and automatically returns 400 if it receives invalid bodies

swsnr committed:
Replace InstanceCreationResult with Either over InstanceError

swsnr committed:
Replace InstanceUpdateResult with Either over InstanceError

swsnr committed:
Turn Role in nice enum

FRosner commented on server/src/main/scala/de/frosner/broccoli/models/InstanceError.scala:
```diff
@@ -0,0 +1,101 @@
+    *
+    * @param throwable The throwable thrown by the operation
+    */
+  @deprecated(message = "Do not use, map throwables to specific instance errors or propagate them to trigger ISEs",
```
What are ISEs?

swsnr replied:
"Internal Server Error", ie, HTTP 500 responses.

FRosner commented on server/src/main/scala/de/frosner/broccoli/models/InstanceError.scala:
```diff
@@ -0,0 +1,101 @@
+    * instance.
+    *
+    * @param instanceId The ID of the instance
+    * @param user The user who tried to access the instance
```
ScalaDoc does not match the parameters

swsnr replied:
Fixed.

FRosner commented on server/src/main/scala/de/frosner/broccoli/controllers/InstanceController.scala:
```diff
@@ -51,54 +52,28 @@ case class InstanceController @Inject()(instanceService: InstanceService, overri
+    InstanceController
+      .create(request.body, loggedIn, instanceService)
+      .fold(
+        _.toHTTPResult,
```
How does Scala know how to convert the error into an HTTP result? I see the ToHTTPResult typeclass but I'm not sure I understand how it works.

swsnr replied:
The `@typeclass` annotation generates boilerplate for type classes (see [README](https://github.com/mpilquist/simulacrum)), including an `ops` object which provides syntax support (hence the "ops" import above).  Essentially the "ops" object turns every method of the type class into an extension method on values that have an instance of the type class available.

And indeed, `InstanceError` has a `ToHTTPResult` instance in its companion object.

swsnr committed:
Disable API documentation to unbreak build

FRosner commented:
You changed `user` to `normaluser`. Thus using the default `application.conf` breaks:

```
17:03:48.246 ERROR SecurityService - Error parsing broccoli.auth.conf.accounts: java.util.NoSuchElementException: user is not a member of Enum (administrator, operator, normaluser)
```

If we make this change we need to document it in the migration guide, adjust the documentation in the wiki and adjust the config as well. To track this I prefer creating a ticket because I am going through the tickets to see if any change in the Wiki is required when releasing.

Is there a strong reason to rename it?

swsnr commented:
It was `NormalUser` before; I just didn't realise that the entry names were different from the enum member names.  Arguably that wasn't the best idea to begin with.  I'll rename to `User`.

Edit: Renamed to `User`.

swsnr committed:
Define ToHTTPResult for instance results

swsnr committed:
Test ToHTTPResult as far as possible

Unfortunately we can't write arbitrary instances for Template and Instance because these case class
do not-trivial validation in their constructors.  This needs to be refactored at some point, but currently
it prevents us from writing property tests for models that directly or indirectly use these case classes.

The ToHTTPResult instances are indirectly tests by the endpoint tests however, which use hand-crafted
input data, so we still have at least some test coverage.

FRosner reviewed (commented):
Please check why the start stop buttons disappear when you are admin.

swsnr commented:
@FRosner I fixed the role issue; there was a "role.toString" at some places, which doesn't work with enumeratum.  The code's using the real Role type now, and doesn't pass roles around as strings.

The UI works now, please review again :)

FRosner commented:
@lunaryorn where did you add this? The last commit still seems to be the same?

swsnr commented:
@FRosner Thanks <3

swsnr commented:
@FRosner I amended the "Turn Role in nice enum" commit.
